csharp/src/Ice/Communicator-LocatorManager.cs

```diff
@@ -677,158 +432,77 @@ public LocatorKey(ILocatorPrx prx)
-        private readonly Dictionary<LocatorKey, LocatorTable> _locatorTableMap = new Dictionary<LocatorKey, LocatorTable>();
-        private readonly bool _backgroundLocatorCacheUpdates;
     }
 
     internal sealed class LocatorTable
     {
-        internal void Clear()
-        {
-            lock (_mutex)
-            {
-                _adapterEndpointsTable.Clear();
-                _objectTable.Clear();
-            }
-        }
-
-        internal IReadOnlyList<Endpoint>? GetAdapterEndpoints(string adapter, int ttl, out bool cached)
-        {
-            if (ttl == 0) // Locator cache disabled.
-            {
-                cached = false;
-                return null;
-            }
+        private readonly ConcurrentDictionary<string, (TimeSpan Time, IReadOnlyList<Endpoint> Endpoints)>
+            _adapterEndpointsTable =
+                new ConcurrentDictionary<string, (TimeSpan Time, IReadOnlyList<Endpoint> Endpoints)>();
 
-            lock (_mutex)
-            {
-                if (_adapterEndpointsTable.TryGetValue(adapter,
-                    out (long Time, IReadOnlyList<Endpoint> Endpoints) entry))
-                {
-                    cached = CheckTTL(entry.Time, ttl);
-                    return entry.Endpoints;
-                }
-                cached = false;
-                return null;
-            }
-        }
+        private readonly ConcurrentDictionary<Identity, (TimeSpan Time, Reference Reference)>
+            _objectReferenceTable =
+                new ConcurrentDictionary<Identity, (TimeSpan Time, Reference Reference)>();
 
-        internal void AddAdapterEndpoints(string adapter, IReadOnlyList<Endpoint> endpoints)
-        {
-            lock (_mutex)
-            {
-                _adapterEndpointsTable[adapter] = (Time.CurrentMonotonicTimeMillis(), endpoints);
-            }
-        }
+        internal void AddAdapterEndpoints(string adapter, IReadOnlyList<Endpoint> endpoints) =>
+            _adapterEndpointsTable[adapter] = (Time.CurrentMonotonicTime(), endpoints);
 
-        internal IReadOnlyList<Endpoint>? RemoveAdapterEndpoints(string adapter)
-        {
-            lock (_mutex)
-            {
-                if (_adapterEndpointsTable.TryGetValue(adapter,
-                    out (long Time, IReadOnlyList<Endpoint> Endpoints) entry))
-                {
-                    _adapterEndpointsTable.Remove(adapter);
-                    return entry.Endpoints;
-                }
-                return null;
-            }
-        }
+        internal void AddObjectReference(Identity id, Reference reference) =>
+            _objectReferenceTable[id] = (Time.CurrentMonotonicTime(), reference);
 
-        internal Reference? GetObjectReference(Identity id, int ttl, out bool cached)
+        internal (IReadOnlyList<Endpoint>?, bool) GetAdapterEndpoints(string adapter, TimeSpan ttl)
```

Comment:
We should use names  for the tuple elements
```suggestion
        internal (IReadOnlyList<Endpoint>? Endpoints, bool Cached) GetAdapterEndpoints(string adapter, TimeSpan ttl)
```

Reply:
Returning stale endpoints is useful to implement [background locator cache updates](https://doc.zeroc.com/ice/3.7/property-reference/miscellaneous-ice-properties#id-.MiscellaneousIce.*Propertiesv3.7-Ice.BackgroundLocatorCacheUpdates) and it's why we can return stale endpoints here.